internal/client/db/backend.go

```diff
@@ -0,0 +1,1835 @@
+// Copyright 2024 ChainSafe Systems (ON)
+// SPDX-License-Identifier: LGPL-3.0-only
+
+package db
+
+import (
+	"errors"
+	"fmt"
+	"maps"
+	"sync"
+
+	"github.com/ChainSafe/gossamer/internal/client/api"
+	"github.com/ChainSafe/gossamer/internal/client/api/utils"
+	"github.com/ChainSafe/gossamer/internal/client/db/columns"
+	"github.com/ChainSafe/gossamer/internal/client/db/metakeys"
+	"github.com/ChainSafe/gossamer/internal/client/db/offchain"
+	statedb "github.com/ChainSafe/gossamer/internal/client/state-db"
+	hashdb "github.com/ChainSafe/gossamer/internal/hash-db"
+	"github.com/ChainSafe/gossamer/internal/log"
+	memorydb "github.com/ChainSafe/gossamer/internal/memory-db"
+	"github.com/ChainSafe/gossamer/internal/primitives/blockchain"
+	"github.com/ChainSafe/gossamer/internal/primitives/core/hash"
+	p_offchain "github.com/ChainSafe/gossamer/internal/primitives/core/offchain"
+	"github.com/ChainSafe/gossamer/internal/primitives/database"
+	"github.com/ChainSafe/gossamer/internal/primitives/runtime"
+	"github.com/ChainSafe/gossamer/internal/primitives/runtime/generic"
+	statemachine "github.com/ChainSafe/gossamer/internal/primitives/state-machine"
+	"github.com/ChainSafe/gossamer/internal/primitives/storage"
+	"github.com/ChainSafe/gossamer/internal/primitives/storage/keys"
+	"github.com/ChainSafe/gossamer/internal/primitives/trie"
+	"github.com/ChainSafe/gossamer/internal/primitives/trie/cache"
+	"github.com/ChainSafe/gossamer/internal/saturating"
+	"github.com/ChainSafe/gossamer/pkg/scale"
+	"github.com/ChainSafe/gossamer/pkg/trie/triedb"
+	"golang.org/x/exp/slices"
+)
+
+var logger = log.NewFromGlobal(log.AddContext("client", "db"))
```

Comment:
usually we use `pkg` as the context so it must be:

```suggestion
var logger = log.NewFromGlobal(log.AddContext("pkg", "db"))
```

Unless we are planning to change our logging strategy

Reply:
I assumed the first was just the parent package.  I think it makes more sense given I want to group these loggers with context key "client".

Reply:
Noup, it is a key-value pair that will be show as `client=db` in the logs.
You could use a combination of `pkg` and `module` as we are [doing here](https://github.com/ChainSafe/gossamer/blob/b6a7ef6ad8f22b9a777948d323eeea29a703ff79/lib/runtime/wazero/imports.go#L32-L35) so if you wanna filter you can filter by `pkg=db` or `module=client` as a second level.
My goal here with this comment is to try to keep our conventions

Reply:
Ended up using a slash `/`in the values and using `pkg` as the key.  I think that's what the author of the logging library intended ([code](https://github.com/ChainSafe/gossamer/blob/604f4d4d2c697fe7403b4705b8a2cfd15399526e/tests/stress/helpers.go#L24)).